db/log_reader.cc

```diff
@@ -450,17 +457,54 @@ unsigned int Reader::ReadPhysicalRecord(Slice* result, size_t* drop_size) {
 
     buffer_.remove_prefix(header_size + length);
 
-    *result = Slice(header + header_size, length);
-    return type;
+    if (uncompress_ && type != kSetCompressionType) {
+      // Uncompress compressed records
+      uncompressed_record_.clear();
+      size_t uncompressed_size = 0;
+      int remaining = 0;
+      do {
+        remaining = uncompress_->Uncompress(header + header_size, length,
+                                            uncompressed_buffer_.get(),
+                                            &uncompressed_size);
+        if (remaining < 0) {
+          buffer_.clear();
+          return kBadRecord;
+        }
+        if (uncompressed_size > 0) {
+          uncompressed_record_.append(uncompressed_buffer_.get(),
+                                      uncompressed_size);
+        }
+      } while (remaining > 0);
```

Comment:
Also, I think your original check of ```remaining > 0 || uncompressed_size == kBlockSize``` was probably correct. The [documentation](http://facebook.github.io/zstd/zstd_manual.html#Chapter9) states that if output.pos == output.size, there may be some data left in internal buffers.

Reply:
I think that's why I had added it, although removing it didn't cause the tests to fail.